rfcs/convergence/event-handlers-arguments.md

```diff
@@ -0,0 +1,170 @@
+
+@dzearing @layershifter
+
+## Background
+
+We should provide guidance on how events handlers (`onChange`, `onOpen`, etc.) are exposed across components. Today we have a discrepancy between `@fluentui/react-northstar` (Fluent UI Northstar) and `@fluentui/react` (Fluent UI v8) components:
+
+```tsx
+// v8
+// "newValue" is the new value of the component's main data, depending on the component type
+onChange: (ev: React.FormEvent, newValue:  ___);
+```
+
+```tsx
+// Northstar
+// "data" is props, with the new value mixed in on top
+// const data = { ...props, value: newValue }
+onChange: (ev: React.FormEvent, data: TProps);
+```
+
+For converged components (Fluent UI vNext) it's implemented in different ways 😪
+
+## Use cases for event handlers
+
+1. The primary data has changed, and I need the new value to support controlled component cases or to forward the data to a state management system.
+
+2. When the data changes I not only need the new value, but I need the original props of the component, so that I can differentiate which instance fired the change event to a general handler.
+
+There are a few types of components which might fire an `onChange` event:
+
+- Single literal value components (`value = "foo"`): `Input` and `Toggle`
+- Single-object value components (`value = { name: 'foo-item' }`): `RadioGroup`, `Select`
+- Multi-object selection components (`value = ["foo", "bar"]`): multiselect `Dropdown`, `DetailsList`, `Tree`
+
+## Problem statement
+
+We need to standardize event handlers across components and ensure that API is consistent to avoid confusion for customers. To clarify, we should avoid this case:
+
+```tsx
+// ❌ Each component in this example has completely different signature for event handlers
+<>
+  <Checkbox onChange={(ev: React.FormEvent, newChecked: boolean) => {}} />
+  <Input onChange={(ev: React.FormEvent, data: { value: string }) => {}} />
+</>
+```
+
+## Prior considerations
+
+Semantic UI React's thread for `onChange` which provided some background on thinking for Option B below:
+https://github.com/Semantic-Org/Semantic-UI-React/issues/623#issuecomment-261018287
+
+Material UI follows a similar approach to v8 code: the second argument represents the new value being changed. See [Select API](https://material-ui.com/api/select/) as an example of this. However the components aren't consistent - See [Input API](https://material-ui.com/api/input/) or [Switch API](https://material-ui.com/api/switch/) where neither have a second argument.
+
+## Detailed Design or Proposal
+
+### Option A (recommended):
+
+- Standardize on a `data` object where we provide the following minimum props:
+
+| Prop    | Description                                      |
+| ------- | ------------------------------------------------ |
+| `value` | new value for the data of the component          |
+| `props` | the original props given to the parent component |
+
+Example:
+
+```tsx
+interface ChangeData<TValue, TProps> {
+  value: TValue;
```

Comment:
The name `value` doesn't always make sense and could possibly be confusing in some cases. E.g. for Checkbox, the `onChange` event happens when the `checked` prop changes, _not_ the `value` prop. It could be more natural if either:
1. It's named `data.checked`, or
2. The event is named `onCheckedChange`

For option 1, this type could be:
```ts
type ChangeData<TPropName extends keyof TProps, TProps> = Pick<TProps, TPropName> & { props: TProps; }
```

The event would be:
```ts
onChange?: (data: ChangeData<'checked', CheckboxProps>) => void;
```

And it'd be used like:
```ts
<Checkbox onChange={data => console.log(data.checked)} />
```

Reply:
> The name `value` doesn't always make sense and could possibly be confusing in some cases.

To be clear, there is no plan to use the same `ChangeData` interface for all components, I will update this section to make clearer. Totally agree that `value` does not have sense for `Checkbox`, especially when `Checkbox` has `checked` prop.

Reply:
@behowell I applied your feedback in a0176c6, is it clearer?